RNA_folding.py

```diff
@@ -0,0 +1,319 @@
+# Copyright 2021 D-Wave Systems
+# Based on the paper 'RNA folding using quantum computers’
+# Fox DM, MacDermaid CM, Schreij AM, Zwierzyna M, Walker RC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from itertools import product, combinations
+import numpy as np
+import networkx as nx
+import matplotlib.pyplot as plt
+import dimod
+from dwave.system import LeapHybridCQMSampler
+
+"""
+This file contains functions called  to help in the building and processing of models related to RNA folding.
+The models are built and solved in a different file.
+"""
+
+
+def text_to_matrix(file_name, min_loop=2):
+    """ Reads properly formatted RNA text file and returns a matrix of possible hydrogen bonding pairs.
+
+    Args:
+        file_name: String path to text file.
+        min_loop: Integer minimum number of nucleotides between two bonding nucleotides.
+
+    Returns: Numpy matrix of 0's and 1's, where 1 represents a possible bonding pair.
+    """
+
+    # Requires text file of rna data written in same format as examples.
+    with open(file_name) as f:
+        rna = "".join(("".join(line.split()[1:]) for line in f.readlines())).lower()
+
+    # Create a dictionary of all indices where each nucleotide occurs.
+    # Mote: t vs u is mainly preference of database. While u is in rna, t is in dna.
+    index_dict = {'a': [], 'c': [], 'g': [], 't': [], 'u': []}
```

Comment:
You might consider using `defaultdict(list)` here (from the `collections` package):
```suggestion
    index_dict = defaultdict(list)
```

Reply:
Good call.